UTRannotator.pm

```diff
@@ -0,0 +1,1306 @@
+                        my @stops = sort {$a <=> $b} @{$ref_existing_oORF{$start_pos}};
+                        $uFrameshift_ref_type_length = $stops[0]-$start_pos+3;
+                    } else {
+                        $uFrameshift_ref_type_length = "NA";
+                    }
+
+                    $uFrameshift_StartDistanceToCDS = $utr_length - $start_pos;
+
+                    #if there is an alternative stop codon in the mutant uORF sequence
+                    my %mut_uORF = %{$self->existing_uORF(\@mut_utr_seq)};
+                    my @alt_overlapping_seq = split //, $mut_utr_seq.$UTR_info->{cds_seq};
+                    my %alt_existing_oORF = %{$self->existing_uORF(\@alt_overlapping_seq)};
+
+                    if (exists($mut_uORF{$start_pos})){
+                    $uFrameshift_alt_type = "uORF";
+                    } elsif(($length-$start_pos)%3){
+                        $uFrameshift_alt_type = "OutOfFrame_oORF";
+                    } else {
+                        $uFrameshift_alt_type = "InFrame_oORF";
+                    }
+
+                    #get the length
+                    if (exists($alt_existing_oORF{$start_pos})){
+                        my @stops = sort {$a <=> $b} @{$alt_existing_oORF{$start_pos}};
+                        $uFrameshift_alt_type_length = $stops[0]-$start_pos+3;
+                    } else {
+                        $uFrameshift_alt_type_length = "NA";
+                    }
+
+                    #find evidence from sorfs.org
+
+                    $uFrameshift_evidence= (exists $self->{uORF_evidence})? $self->find_uorf_evidence($UTR_info,$chr,$start_pos): "NA";
+
+                    my %uORF_effect = (
+                        "uFrameShift_ref_type" => $uFrameshift_ref_type,
+                        "uFrameShift_ref_type_length" => $uFrameshift_ref_type_length,
+                        "uFrameShift_ref_StartDistanceToCDS" => $uFrameshift_StartDistanceToCDS,
+                        "uFrameShift_alt_type" => $uFrameshift_alt_type,
+                        "uFrameShift_alt_type_length" => $uFrameshift_alt_type_length,
+                        "uFrameShift_KozakContext" => $uFrameshift_KozakContext,
+                        "uFrameShift_KozakStrength" => $uFrameshift_KozakStrength,
+                        "uFrameShift_Evidence" => $uFrameshift_evidence,
+                    );
+
+                    $output_flag = $output_flag? $output_flag."&"."uFrameShift":"uFrameShift";
+                    $output_effects = $output_effects? $output_effects."&".$self->transform_hash_to_string(\%uORF_effect):$self->transform_hash_to_string(\%uORF_effect);
+                }
+
+        }
+    }
+
+    $result{'uFrameShift_flag'} = $output_flag;
+    $result{'uFrameShift_effect'} = $output_effects;
+
+    return \%result;
+}
+
+#########
+# Utils #
+#########
+
+sub count_number_ATG {
+
+    #Description: count the number of existing ATGs in the five prime UTR sequence
```

Comment:
The details could be before the method. For example:

```
=head2 count_number_ATG

  Arg [1]    : $seq
  Description: Count the number of existing ATGs in the five prime UTR sequence.
  Return: returns the number of uORFs and the number of oORFs
  Returntype : listref

=cut
```
It returns a reference to a hash?

Reply:
Corrected